src/clinical/clinical-service.ts

```diff
@@ -128,3 +134,68 @@ export const updateDonorStats = async (donorId: number, coreCompletionOverride:
 
   return await donorDao.update(updatedDonor);
 };
+
+export async function getClinicalData(programId: string) {
+  if (!programId) throw new Error('Missing programId!');
+
+  const start = new Date().getTime() / 1000;
+
+  const donors = await getDonors(programId);
+
+  // collect all records
+  const recordsMap: any = {};
+  donors.forEach(d => {
+    Object.values(ClinicalEntitySchemaNames).forEach(entity => {
+      let clincialInfoRecords;
+      if (entity === ClinicalEntitySchemaNames.REGISTRATION) {
+        clincialInfoRecords = generateSampleRegistrationDataForDonor(d);
+      } else {
+        clincialInfoRecords = getClinicalEntitiesFromDonorBySchemaName(d, entity);
+      }
+      recordsMap[entity] = _.concat(recordsMap[entity] || [], clincialInfoRecords);
+    });
+  });
+
+  // map into object ready for api processing
+  const schemasWithFields = dictionaryManager.instance().getSchemasWithFields();
+  const data = Object.entries(recordsMap)
+    .map(([entityName, records]) => {
+      if (isEmpty(records)) return undefined;
+
+      const relevantSchemaWithFields = schemasWithFields.find(s => s.name === entityName);
+
+      return {
+        entityName,
+        records,
+        entityFields: relevantSchemaWithFields?.fields || [],
+      };
+    })
+    .filter(notEmpty);
+
+  const end = new Date().getTime() / 1000;
+  L.info(`getClinicalData took ${end - start}s`);
```

Comment:
so how much time did it take for a bulk of donors ?

Reply:
this can also be debug statement not info

Reply:
Running locally here's what I'm getting:
- 20,000 donors with sample registration take about 6 secs
- 3000 donors with sample registration is about 1 sec

I added donor clinical info in the 3000 case and it still took about 1 sec.

Reply:
Can't speak for dev/qa until I run their. Also the zip file sizes aren't too big compared, its in ~KB range.

Reply:
so it's blocking for 1 second at 3k

Reply:
yea, looks like it.

Reply:
my suggestion is to do this in a thread pool if time permits or open a ticket to convert it to thread pool because we don't know how much clinical data it will add, it's the right way to do it